python/interpret-core/interpret/blackbox/permutationimportance.py

```diff
@@ -0,0 +1,210 @@
+        if sp.sparse.issparse(shuffled_prediction):
+            shuffled_prediction = shuffled_prediction.toarray()
+        metric = self.metric(self.labels, shuffled_prediction)
+        importance_score = base_metric - metric
+        global_importance_values[idx] = importance_score
+
+    def _compute_sparse_metric(self, dataset, col_idx, random_indexes, shuffled_dataset,
+                               predict_function, true_labels, base_metric, global_importance_values):
+        # Get non zero column indexes
+        indptr = dataset.indptr
+        indices = dataset.indices
+        col_nz_indices = indices[indptr[col_idx]:indptr[col_idx + 1]]
+        # Sparse optimization: If all zeros, skip the column!  Shuffling won't make a difference to metric.
+        if col_nz_indices.size == 0:
+            return
+        data = dataset.data
+        # Replace non-zero indexes with shuffled indexes
+        col_random_indexes = random_indexes[0:len(col_nz_indices)]
+        with warnings.catch_warnings():
+            warnings.simplefilter('ignore', SparseEfficiencyWarning)
+            # Shuffle the sparse column indexes
+            shuffled_dataset[col_random_indexes, col_idx] = shuffled_dataset[col_nz_indices, col_idx].T
+            # Get set difference and zero-out indexes that had a value but now should be zero
+            difference_nz_random = list(set(col_nz_indices).difference(set(col_random_indexes)))
+            difference_random_nz = list(set(col_random_indexes).difference(set(col_nz_indices)))
+            # Set values that should not be sparse explicitly to zeros
+            shuffled_dataset[difference_nz_random, col_idx] = np.zeros((len(difference_nz_random)),
+                                                                       dtype=data.dtype)
+            idx = col_idx
+            self._add_metric(predict_function, shuffled_dataset, true_labels,
+                             base_metric, global_importance_values, idx)
+            # Restore column back to previous state by undoing shuffle
+            shuffled_dataset[col_nz_indices, col_idx] = shuffled_dataset[col_random_indexes, col_idx].T
+            shuffled_dataset[difference_random_nz, col_idx] = np.zeros((len(difference_random_nz)),
+                                                                       dtype=data.dtype)
+
+    def _compute_dense_metric(self, dataset, col_idx, subset_idx, random_indexes,
+                              predict_function, true_labels, base_metric, global_importance_values):
+        # Create a copy of the original dataset
+        shuffled_dataset = np.array(dataset, copy=True)
+        # Shuffle one of the columns in place
+        shuffled_dataset[:, col_idx] = shuffled_dataset[random_indexes, col_idx]
+        idx = col_idx
+        self._add_metric(predict_function, shuffled_dataset, true_labels,
+                         base_metric, global_importance_values, idx)
+
+    def explain_global(self, name=None):
```

Comment:
It's very weird to me that data goes in the constructor, effectively making this a one-time use object. Why is that the case?

Reply:
This is the convention within the repo. I do think this should be looked into with a design review for validation with external repos. However, for this PR I believe keeping to convention is the more correct approach.


Long winded answer :
When compared to glassbox models, it is consistent, since fit is only called once. However, you are right, creating two global explanations for comparison requires two constructor calls and two explain_global calls which seems off.

My initial implementation had explain_global(X, y) 
# similar to ebm.fit(X, y) and *.explain_local(X, y)

In general I agree that a convention should be chosen. I prefer only constructor or only explain_global. While reviewing explainers for consistency in both repos I saw a few interpret-community explainers with both constructor data and explain_global data requirements in a subset of explainers. That really confused me.

My preference after the review tended toward explain_global(X, y, name=None), with no data passed to any constructor. It felt more in line with explain_local(X, y) and fit(X, y) APIs found within interpret. It also could help with creating an Explainer, serializing it, and then explaining global on datasets on different machines with different data.

Either way, I think keeping to the repo convention is the correct approach for this PR. And a github issue for reviewing this convention would be worthwhile, even to just gather supporting evidence for the convention chosen.

Reply:
Sure, consistency for this PR makes sense. The constructor vs. explain_global data in interpret-community is train vs. test data, at least as far as I'm aware. We don't require an explicit `fit` call from the user, but we assume that they'll want training data to be consistent while test data will vary.

Reply:
I see. So you can call explain global on different test sets with the init on training data. 

That makes some sense. Definitely worth bringing up in the discussion